lib/mongoid/clients/sessions.rb

```diff
@@ -51,6 +51,67 @@ def with_session(options = {})
         Threaded.clear_session
       end
 
+      # Execute a block within the context of a transaction. Note that this can only occur within
+      # a session.
+      #
+      # @example Execute some operations in a transaction.
+      #   band.with_session do
+      #     band.with_transaction do
```

Comment:
What is preventing Mongoid from automatically executing `with_session` when a user wants a transaction?

Reply:
The issue is that sessions can't get nested. If it automatically calls `with_session`, then without extra logic it would raise an error if it were called from within a session block. I'm not big on the idea of having `with_transaction` only sometimes implicitly start an explicit session; a better idea IMO would be to have a separate method for starting a transaction and a session at once (e.g. `with_session_and_transaction`), but I'm not super fond of the cognitive overhead of forcing users to choose between which method to call depending on the situation. I'd be interested to hear your thoughts on the tradeoffs between having `with_transaction`'s behavior be dependent on the context, having a separate method for starting a transaction and a session at once, or just having the user have to call both methods in the case that a session isn't already started.

Reply:
I would do a github code search for `with_session` to see if there are any instances of people actually using that method.

https://docs.mongodb.com/manual/reference/server-sessions/ says:

> MongoDB’s server sessions, or logical sessions, are the underlying framework used by client sessions to support Causal Consistency and retryable writes.

I would expect that users of Mongoid, being a higher level ODM, would most likely prefer to have implicit sessions created for them all the time, and hence would almost never call `with_session` themselves.

Reply:
I'm not sure I accept the premise that a Github code search would give a useful sample of projects; I'd imagine that most real-world applications using Mongoid are not open source.

Reply:
My impression of the session concept is the entire application should generally be using sessions all the time, or not. Hence I'm thinking whether sessions are used by default should be configurable in mongoid.yml and equivalents.

Is there a performance penalty for having sessions on all the time, and is it quantified anywhere?

Reply:
That's a good question! I know that the use of sessions is to facilitate things like transactions, causally consistent reads, and retryable writes. I'm not sure if the general usage pattern is to enable them for the entire lifetime of a client or just sections. The spec doesn't mention anything about a URI option for a session (which is how most things that last the entire lifetime of the client are configured), but that might just be because there isn't any good way to get back that session when creating a client from a URI. Let's check around and see what others on the drivers team think about this.

Reply:
Per offline discussion, we'll leave this as-is for now and open a new ticket to make sessions used by default on all Mongoid clients.

Reply:
This PR depends on the implicit sessions as discussed.